BlueSky: Token expired

socialmetricbot | [2023-10-08T11:17:55.156] [ERROR] i.g.b.s.m.w.b.w.BskyMentionListener - could not retrieve mentions: [java.lang.IllegalStateException: Getting notifications not successful. RC=400. Body: >>{"error":"ExpiredToken","message":"Token has expired"}<<.].
socialmetricbot | java.util.concurrent.CompletionException: java.lang.IllegalStateException: Getting notifications not successful. RC=400. Body: >>{"error":"ExpiredToken","message":"Token has expired"}<<.

--- a/bsky-client/src/main/java/io/github/bmarwell/social/metricbot/bsky/DefaultBlueSkyClient.java
+++ b/bsky-client/src/main/java/io/github/bmarwell/social/metricbot/bsky/DefaultBlueSkyClient.java
@@ -50,7 +50,7 @@ public class DefaultBlueSkyClient implements BlueSkyClient {
         if (isLoggedIn) {
             LOG.debug("[BSKY] Already logged in.");
 
-            if (Instant.now().isAfter(this.refreshBefore.minusSeconds(60))) {
+            if (isTokenExpired()) {
                 return doLoginAndSetVariables();
             }
 
@@ -65,6 +65,10 @@ public class DefaultBlueSkyClient implements BlueSkyClient {
         return doLoginAndSetVariables();
     }
 
+    protected boolean isTokenExpired() {
+        return Instant.now().isAfter(this.refreshBefore.minusSeconds(60));
+    }
+
     private CompletableFuture<Void> doLoginAndSetVariables() {
         return CompletableFuture.supplyAsync(this::doLogin)
                 .handle((final Optional<AtProtoLoginResponse> result, final Throwable error) -> {
@@ -74,6 +78,7 @@ public class DefaultBlueSkyClient implements BlueSkyClient {
                         this.accessToken.set("");
                         this.isLoggedIn = false;
                         this.loginNotSuccessfull = true;
+                        this.refreshBefore = Instant.now().plusSeconds(360);
                         return null;
                     }
 
@@ -82,6 +87,7 @@ public class DefaultBlueSkyClient implements BlueSkyClient {
                     this.accessToken.set(result.orElseThrow().accessJwt());
                     this.isLoggedIn = true;
                     this.loginNotSuccessfull = false;
+                    this.refreshBefore = Instant.now().plusSeconds(360);
                     return null;
                 });
     }
@@ -302,6 +308,18 @@ public class DefaultBlueSkyClient implements BlueSkyClient {
 
     public Client getClient() {
         return client;
+    }
+
+    protected AtomicReference<String> getAccessToken() {
+        return accessToken;
+    }
+
+    protected Instant getRefreshBefore() {
+        return refreshBefore;
+    }
+
+    protected void setRefreshBefore(final Instant refreshBefore) {
+        this.refreshBefore = refreshBefore;
     }
 
     @Override

--- a/bsky-client/src/test/java/io/github/bmarwell/social/metricbot/bsky/DefaultBlueSkyClientTest.java
+++ b/bsky-client/src/test/java/io/github/bmarwell/social/metricbot/bsky/DefaultBlueSkyClientTest.java
@@ -37,4 +37,17 @@ class DefaultBlueSkyClientTest {
         // then:
         assertThat(link).hasPath("/profile/metricbot.de/post/3kb4fxhzeed25");
     }
+
+    @Test
+    void token_expired_detected() {
+        // given
+        final var client1 = new DefaultBlueSkyClient(this.bsc);
+        client1.setRefreshBefore(Instant.now().minusSeconds(600));
+
+        // when
+        final var tokenExpired = client1.isTokenExpired();
+
+        // then
+        assertThat(tokenExpired).isTrue();
+    }
 }

